polar-core/src/debugger.rs

```diff
@@ -229,18 +207,97 @@ impl Debugger {
                     ),
                 });
             }
-            "n" | "next" | "over" => {
-                self.step = Some(Step::Over {
-                    snapshot: vm.queries[..vm.queries.len().saturating_sub(1)].to_vec(),
-                })
+            "query" => {
+                let mut level = parts.get(1).and_then(|s| s.parse().ok()).unwrap_or(0);
+                let mut trace_stack = vm.trace_stack.clone();
+
+                // Walk up the trace stack to get the query at the requested level.
+                let mut term = vm.trace.last().and_then(|t| t.term());
+                while level > 0 {
+                    if let Some(trace) = trace_stack.pop().map(|ts| ts.as_ref().clone()) {
+                        if let Some(t) = trace.last() {
+                            if let Trace{node: Node::Term(t), ..} = &**t {
+                                term = Some(t.clone());
+                                level-=1;
+                            }
+                        }
+                    } else {
+                        return Some(Goal::Debug {
+                            message: "Error: level is out of range".to_owned()
+                        })
+                    }
+                }
+
+                if let Some(query) = term {
+                    return Some(Goal::Debug {
+                        message: vm.query_summary(&query)});
+                } else {
+                    return Some(Goal::Debug {
+                        message: "".to_owned()
+                    })
+                }
             }
-            "out" => {
-                self.step = Some(Step::Out {
-                    snapshot: vm.queries[..vm.queries.len().saturating_sub(3)].to_vec(),
+            "stack" | "trace" => {
```

Comment:
nit: same as above, this is a very long match arm that might be better split out into a separate function